include/seqan3/search/detail/search_configurator.hpp

```diff
@@ -100,26 +150,87 @@ class search_configurator
-        {
-            using algorithm_t = search_scheme_algorithm<configuration_t,
-                                                        index_t,
-                                                        policy_max_error,
-                                                        policy_search_result_builder<configuration_t>>;
-            return algorithm_t{cfg, index};
-        }
-        else
-        {
-            using algorithm_t = unidirectional_search_algorithm<configuration_t,
-                                                                index_t,
-                                                                policy_max_error,
-                                                                policy_search_result_builder<configuration_t>>;
-            return algorithm_t{cfg, index};
-        }
+        using search_result_t = typename select_search_result<configuration_t, index_t>::type;
+        using search_result_collection_t = std::vector<search_result_t>;
+        using type_erased_algorithm_t = std::function<search_result_collection_t(query_t)>;
+
+        return configure_hit_strategy<type_erased_algorithm_t>(cfg | search_cfg::detail::result_type<search_result_t>,
+                                                               index);
     }
+
+    template <typename algorithm_t, typename configuration_t, typename index_t>
+    static algorithm_t configure_hit_strategy(configuration_t const &, index_t const &);
 };
 
+/*!\brief Configures the algorithm with the correct hit strategy.
+ *
+ * \tparam algorithm_t The type erased algorithm used for the fixed return type.
+ * \tparam configuration_t The type of the search configuration.
+ * \tparam index_t The type of the index.
+ *
+ * \returns The configured search algorithm.
+ *
+ * \details
+ *
+ * If the algorithm was configured with the dynamic hit configuration element seqan3::search_cfg::hit, the
+ * configuration element is removed and replaced by the selected static hit configuration element.
+ * If the hit configuration element is already a static one nothing is changed in the configuration.
+ * After selecting the correct hit strategy the corresponding search algorithm is created with the new configuration
+ * and the given index.
+ *
+ * \throws std::invalid_argument if the dynamic hit configuration was not initialised with a hit strategy.
+ *
+ * If no hit configuration was configured a static assert is emitted during compilation.
+ */
+template <typename algorithm_t, typename configuration_t, typename index_t>
+algorithm_t search_configurator::configure_hit_strategy(configuration_t const & cfg, index_t const & index)
+{
+    // Select the actual search algorithm and return an instance constructed with the adapted config and the index.
+    auto generate_algorithm = [&] (auto new_cfg) -> algorithm_t
+    {
+        using new_configuration_t = decltype(new_cfg);
+        using selected_algorithm_t =
+            typename select_search_algorithm<new_configuration_t,
```

Comment:
It is a little unintuitive that the function `configure_hit_strategy` also selects the search algorithm. Can this be done outside of this function?

Reply:
I could make an explicit function to just handle the algorithm selection. The idea is to have  a continuation interface: like configure a -> configure b -> configure c -> .... But I also often received the feedback that for 2 lines of code an additional function bloat is not seen. BUt I also agree, that this way we can document the behavior much easier and are flexible with the configuration chain. Then again it keeps the changes minimal and we don't add a feature that we don't neeed yet. I really don't care. If I see for both points pros and cons. Just let me know what you'd prefer.

Reply:
I separated it now in the last fixup commit